File: 56. Merge Intervals/56. Merge Intervals.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function (intervals) {
  intervals.sort((a, b) => a[0] - b[0]);
  const merged = [];

  for (const interval of intervals) {
    n = merged.length;
    if (n === 0 || merged[n - 1][1] < interval[0]) {
      merged.push(interval);
    } else {
      merged[n - 1][1] = Math.max(merged[n - 1][1], interval[1]);
    }
  }

  return merged;
};

File: 56. Merge Intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort()
        merged = []

        for l, h in intervals:
            if not merged or merged[-1][1] < l:
                merged.append([l, h])
            else:
                merged[-1][1] = max(merged[-1][1], h)

        return merged
